oshi-core/src/main/java/oshi/util/Memoizer.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * OSHI (https://github.com/oshi/oshi)
+ *
+ * Copyright (c) 2010 - 2019 The OSHI Project Team:
+ * https://github.com/oshi/oshi/graphs/contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package oshi.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A memoized function stores the output corresponding to some set of specific
+ * inputs. Subsequent calls with remembered inputs return the remembered result
+ * rather than recalculating it.
+ */
+public final class Memoizer {
+
+    private Memoizer() {
+    }
+
+    /**
+     * Store the function in a thread-safe map to be computed only once.
+     * 
+     * @param f
+     *            The {@link java.util.function.Function} to memoize
+     * @return A memoized version of the function
+     */
+    public static <I, O> Function<I, O> memoize(Function<I, O> f) {
```

Comment:
While implementing lazy-initialization with CHM is short, it is an overkill at the very least on terms of memory footprint. It's possible to implement `memoize(Function)` with DCL, and then implement `memoize(Supplier)` by simply reusing `memoize(Function)` (because `Supplier` is essentially a `Function` that ignores its argument):
```java
public static <I, O> Function<I, O> memoize(Function<I, O> f) {
  return new Function<I, O>() {
    private volatile O value;

    public O apply(I arg) {
      O value = this.value;
      if (value == null) {
        synchronized (this) {
          value = this.value;
          if (value == null) {
            this.value = value = f.apply(i);
            if (value == null) {
              throw new AssertionError("Internal error. Function must not return null");
            }
          }
        }
      }
      return value;
    }
  };
}

public static <T> Supplier<T> memoize(Supplier<T> s) {
  return () -> memoize(i -> s.get()).apply(null);
}
```

Reply:
Thanks. I might try this later.  I haven't used the function version yet and may not see a need to.  The supplier code below seems rather efficient as well.

Reply:
Also I need to figure out how to implement expiration. Currently only one real use case in the code, reading swapfile stats.  I read used and total at the same time and if the getUsed and getTotal are called "close together" not require a second read.

Reply:
> Also I need to figure out how to implement expiration

As I mentioned, both Guava and my [`MemoizedObject`](https://gist.github.com/stIncMale/2fce5f39aa37ac4584e0635ca6f67fe5) implement expiration, and both contain concurrency tests. I wrote `MemoizedObject` before I discovered it's present in Guava (last time I checked, the implementation were similar).

There are though memory footprint optimizations one may add to `MemoizedObject`:
- use the monitor associated with `this` instead of having a dedicated `monitor ` object;
- use a more simple implementation for situations when there is no expiration to save space by not having fields related to TTL;
- a pair of fields `ttlNanos`, `lastRefreshedTNanos` may potentially be replaced with a single field, but I am not completely sure and this is may not be very straightforward.

Reply:
@stIncMale -- I just tried your DCL version of the memoizer.. it doesn't work (run the main method in the latest commit) .... yes I know you have suggested to use the MemoizedObject, I'm trying to do as much "roll my own" as I can here...